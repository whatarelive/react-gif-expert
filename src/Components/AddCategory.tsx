// Librerias de React.
import React, {useState} from "react";
// Types.
import {AddCategoryProps} from "../Types/types";


export const AddCategory = ( { onNewCategory }: AddCategoryProps ) => {

    // Hook useState : permite dal un estado al texto del input.
    const [ inputValue, setInputValue ] = useState( String );

    // Funcion para extraer el texto del input y guardarlo en el useState.
    const onInputChange = ({ target }: React.ChangeEvent<HTMLInputElement>) => setInputValue( target.value );

    // Funcion que enviar el valor de la nueva categoria al componente padre.
    const onSubmit = (event: React.FormEvent) => {
        event.preventDefault();

        const text = inputValue.trim();
        
        if( text.length > 0 ) {
            onNewCategory( text );
            setInputValue('');
        }
    };

    // Renderizado del componente.
    return (
        <form onSubmit={ onSubmit }>
            <input
                type="text"
                placeholder="Buscar Gifs"
                value={ inputValue }
                onChange={ onInputChange }
            />
        </form>
    );
};