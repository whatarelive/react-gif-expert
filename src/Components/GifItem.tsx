import { GifType } from "../Types/types"

export const GifItem = ( image: GifType ) => {

    return (
         <div className="card-item">          
            <img src={ image.url } alt={ image.title } /> 
            <p>{ image.title }</p>
        </div>      
    )
}
