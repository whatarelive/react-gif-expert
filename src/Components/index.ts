// Archivo de barril para todas las importaciones de la carpeta componentes.
export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem';