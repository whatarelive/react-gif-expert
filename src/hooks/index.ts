export * from './useCategory';
export * from './useFecthGifs';