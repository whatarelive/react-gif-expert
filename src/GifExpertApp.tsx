// Librerias de React.
import {useState} from "react";
// Componentes hijos.
import { AddCategory, GifGrid } from "./Components";


export const GifExpertApp = () => {

    // Estado que contiene el listado de las categorias de los gifs.
    const [ categories, setCategories ] = useState(['One Punch']);

    // Funcion que agrega una nueva categoria al estado inicial.
    const onAddCategory = ( newCategory: string ) => {
        
        // Validacion para saber si la nueva categoria no se encuentra del useState.
        if ( !categories.includes( newCategory ) ) {
            setCategories([ newCategory, ...categories ]);
        }
    }

    // Funcion que crea el listado de elementos.
    const listItem = () => {
        return categories.map( ( category: string ) => {
           return <GifGrid key={ category } category={ category }/>;
        });
    }

    return (
        <>
            <h3>Gif Expert App</h3>
            <AddCategory onNewCategory={ onAddCategory }/>
            <ol>{ listItem() }</ol>
        </>
    );
};