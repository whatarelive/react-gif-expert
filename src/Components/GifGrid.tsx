// Custom Hook
import { useFecthGifs } from "../hooks/useFecthGifs";
// Componetes
import { GifItem } from "./index";
// Types.
import {GifGridProps} from "../Types/types";


export const GifGrid = ({ category }: GifGridProps ) => {

    const { images, isLoading } = useFecthGifs( category );
    
    // Renderizado del componente.
    return (
        <>
            <h3>{ category }</h3>

            {
                isLoading && ( <h2>Cargando.....</h2>)
            }

            <div className="card-grid"> 
            { 
                images.map( images => <GifItem key={ images.id } { ...images }/>)
            }
            </div> 
        </>
    );
};
