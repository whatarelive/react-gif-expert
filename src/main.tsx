// Librerias de React.
import React from 'react'
import ReactDOM from 'react-dom/client'
// Estilos CSS. 
import "./Styles/index.css";
// Componentes propios. 
import {GifExpertApp} from "./GifExpertApp.tsx";

const root = document.getElementById('root');

if (root instanceof HTMLDivElement) {

    ReactDOM.createRoot(root).render(
        <React.StrictMode>
                <GifExpertApp/>
        </React.StrictMode>
    );
} 