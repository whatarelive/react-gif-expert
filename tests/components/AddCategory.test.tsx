import React from "react";
import { fireEvent, render, screen } from "@testing-library/react"
import { AddCategory } from "../../src/Components"

describe('Pruebas en el componente <AddCategory/>', () => { 

    // Objeto auxiliar.
    const target = { value: 'Saitama' };

    // Test #1: Sobre la UI del Componenete.
    test('Snapshot del componente', () => { 
   
        // Se guarda el contenido del componente renderizado. 
        const { container } = render(<AddCategory onNewCategory={ jest.fn() }/>);

        // Se crea una captura del elemento renderiazado.
        expect( container ).toMatchSnapshot();
     })

     // Test #2:  Sobre e Input: Simular la funcion onChange
     test('Debe cambiar el valor del TextInput', () => { 
         
        // Renderizamos el componente
        render(<AddCategory onNewCategory={ jest.fn() }/>);
        
        // Extraemos el input del componente renderizado.
        const input = screen.getByRole('textbox') as HTMLInputElement;

        // Disparamos el evento en el HTMLInputElement.
        fireEvent.input( input, { target });

        // El valor ingresado en el input debe ser igual al esperado.
        expect( input.value ).toBe( 'Saitama' );
     })

     // Test #3: Sobre el Formulario: Simular la funcion onSumit con un valor.
     test('deber llamar onNewCategory si el input tiene un valor', () => {
        // Creamos una funcion usando un Mock de Jest. 
        const onNewCategory = jest.fn();  

        // Renderizamos el componente
        render( <AddCategory onNewCategory={ onNewCategory }/>);

        // Extraemos el input y el formulario del componente renderizado.
        const input = screen.getByRole('textbox') as HTMLInputElement;
        const form = screen.getByRole('form') as HTMLFormElement;

        // Disparamos el evento en el HTMLInputElement
        fireEvent.input( input, { target } );
        
        //Disparamos el evento onSubmit en el Formulario.
        fireEvent.submit( form );

        // Esperamos que el valor del input sea igual a un String vacio.
        expect( input.value ).toBe( '' );
        // Esperamos que la funcion halla sido llamada.
        expect( onNewCategory ).toHaveBeenCalled();
        // Esperamos que la funcion solo sea llamada una sola vez.
        expect( onNewCategory ).toHaveBeenCalledTimes( 1 );
        // Esperamos que la funcion sea llamada con el valor de la caja de texto.
        expect( onNewCategory ).toHaveBeenCalledWith( target.value );
      })

      // Test #4: Sobre el Formulario: Simular la funcion onSubmit sin tener un valor.
      test('no debe llamar el onNewCategory si el input esta vacio', () => {

         // Se crea una funcion auxiliar usando un Mock de Jest.
         const onNewCategory = jest.fn();
         
         // Renderizamos el componente
         render(<AddCategory onNewCategory={ onNewCategory }/>)

         // Extraemos el formulario del componenete renderizado.
         const form = screen.getByRole('form') as HTMLFormElement;

         // Disparamos el evento onSubmit en el Formulario.
         fireEvent.submit( form );

         // Esperamos que la funcion no halla sido llamada ni una vez;
         //expect( onNewCategory ).toHaveBeenCalledTimes( 0 );
         expect( onNewCategory ).not.toHaveBeenCalled();
      })
 })