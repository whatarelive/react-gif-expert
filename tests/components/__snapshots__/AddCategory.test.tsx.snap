// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <AddCategory/> Snapshot del componente 1`] = `
<div>
  <form
    aria-label="form"
  >
    <input
      placeholder="Buscar Gifs"
      type="text"
      value=""
    />
  </form>
</div>
`;
